pkg/file/file.go

```diff
@@ -90,11 +96,10 @@ func AtomicWrite(path string, data []byte, mode os.FileMode) (err error) {
 
 	_, err = tmpFile.Write(data)
 	if err != nil {
-		if closeErr := tmpFile.Close(); closeErr != nil {
-			err = fmt.Errorf("%s: %w", closeErr.Error(), err)
-		}
 		return
 	}
+
+	needClose = false
```

Comment:
If I understand the logic correctly then you can remove the `needClose` and the close at line 103 and just close it in the deferred func (and also print the error in case of failed close).

Reply:
Not so. Using os.Rename for file renaming without closing the file handle may lead to some undefined behavior, including file system corruption, data loss, file handle leakage, and other problems.

Reply:
But it's closed before the rename function is called explicitly on line 98 of the original file.